### ISSUE TITLE
PVR Search and Series Links

### ISSUE THREAD

**Issue by heroineinabook** [Ascoware/get-iplayer-automator#113]
I haven't used GiA in a few months (I've upgraded to 1.10.0) but in the past if I put in the Series-Link "Strike BBC4," It would pick up any show with the word Strike in it (Strike: Career of Evil, Strike: The Silkworm, etc) on BBC4 and add it to the download queue. When I tried it on the new version, since I knew a new series was coming out, it picked up nothing. But when I did a regular search for "Strike," the show came up. So now I have it listed in my Series-Link as "Strike: Career of Evil" in my PVR.

Having a general term picked up, such as "Strike BBC4" is helpful when shows, like Strike, change their series name or add an addendum to their title. Especially when you don't know when a new show is coming out.

Is this something that can be ironed out?

**Comment by jchhenderson** [comment]
I am pretty sure the new episodes are on BBC 1.  Try removing the "4" from your search

**Comment by sleaux-meaux** [comment]
Neither "BBC4" or "BBC1" will work. It's "BBC Four" and "BBC One".  Use what you see in the search results for "Strike". In this case, you could use "*" as the TV Network value to cover all channels. It depends on whether you need to screen out programmes with "Strike" in the name from other channels.

**Comment by sleaux-meaux** [comment]
Also: You can double-click to edit the Series-Link entry you already created for "Strike: Career of Evil" and change it to the desired form.

**Comment by heroineinabook** [comment]
When you do a search in the PVR, it automatically tells you what channel it's in, which is "BBC Four" not "BBC 4." 

Here's an example of what I mean:

I go to the PVR and type in "Strike." Up pops all the listings of the word "Strike" including "Strike: Career of Evil." I add that to Series Link. I know, however, that the Strike series has different names such as "Strike: Career of Evil" and "Strike: The Cuckoo's Calling" are examples.  I also know that a future series is coming up but I don't know the exact title but I want to make sure I can grab it. 

In the past, you could add just "Strike" in the Series LInk and it would catch ANYTHING with the word "Strike" in the title. It didn't need to be specific.

In short: I don't necessarily want to catch specific titles but general keywords. You could do this in previous versions of GiA and now you can't.

My apologies if this was not clear enough.

**Comment by heroineinabook** [comment]
Second in short: When I modify the title in Seris Link from "Strike: Career of Evil" to "Strike," it doesn't work.

**Comment by WilliamOfGaunt** [comment]
Is the comma after “Strike” included in your Series link? If so, it won’t work. Just “Strike” suffices. It’s what I have, and it works.

**Comment by sleaux-meaux** [comment]
> In the past, you could add just "Strike" in the Series LInk and it would catch ANYTHING with the word "Strike" in the title. It didn't need to be specific.

Works fine for me both when I enter it directly or when I edit an existing entry. I have used this functionality for a long time, and nothing has changed in the code afaict. If you have already downloaded all the episodes of Strike, none will be added to the queue by the Series-Link entry, so perhaps that is what you see.

**Comment by heroineinabook** [comment]
Sorry, it's been a long day and I am not trying to be obstinate, just trying to troubleshoot on my side.

I'll check some other shows to see if I'm getting the same issue because it could just be a one-off (which I should have done before posting here, but ah well).

Thanks for the suggestions and help!

**Comment by gacres2** [comment]
@heroineinabook In the PVR you can double-click both the title and the channel. So you can start from "Strike: Career of Evil" and "BBC One", but change that to "Strike:" on any channel (by completely deleting "BBC One"), or even "Strike:" and "BBC".

The PVR has always been quite powerful, but it does fail on shows that have a very simple or common name (for example the crime drama "River" finds lots of other programs too).

**Comment by PhilipRoy** [comment]
In a similar vein....I subscribed to BBC Click (just shows "Click" in the PVR, but it insists on downloading "Click - Short Edition" versions also (an abridged version of each episode) which I would like to exclude. Is there any way to automate this?

**Comment by sleaux-meaux** [comment]
You can change the Show value in your PVR entry to Click$ to force matching "Click" at the end of the programme name. If you don't know why that works, google for information on regular expressions. This is only a partial solution, however, since any programme name ending in "Click" will match. If you already know about regular expressions, you'll know that the pattern should be ^Click$ to force matching a programme name that consists only of the string "Click". However, GiA rather stupidly escapes the circumflex, which makes the search fail.

**Comment by skovatch** [comment]
I have no idea why the regex characters were escaped. NSTask doesn't need it; you aren't going through a shell. I can remove that pretty easily.

**Comment by sleaux-meaux** [comment]
You do want to escape them (including $) when they are added to the series-link list - just not when they are read from the list to be passed to get_iplayer search. In fact, it might be a good idea to add them the way the get_iplayer web pvr does, with regex chars escaped and surrounded by ^ and $ to force an exact programme name match. That's what most people would expect as default behaviour. OTOH, surfacing regular expressions may frighten the horses.

**Comment by PhilipRoy** [comment]
Yep, we horses frighten easily.

Thanks for the info re the use of the $. I know it may not be read by some, but maybe update the help info in the PVR sections with some tips like regular expressions?

**Comment by jchhenderson** [comment]
did not realise that GiA accepted regex - useful info (shame about the caret escaping).

...perhaps the view can be swapped from "normal" to "advanced" to show/hide the regex as per @sleaux-meaux ?

**Comment by sleaux-meaux** [comment]
Upon reflection, when adding series-link items to queue it is probably good enough to keep the status quo except: a) don't escape the circumflex character if at the beginning of name string; b) escape dollar sign along with other regex chars unless it is at the end of the name string. 

In practice, regular expressions are not very useful in the PVR. The few cases where they are actually helpful can be accomplished with just circumflex and dollar sign to anchor matches to start/end, but GiA could deal with those behind the scenes automatically:

1. Use an exact match to avoid programme variants, e.g., using `^Click$` to avoid "Click - Short Edition" (if circumflex wasn't escaped in GiA). Such programmes are very rare.  GiA could automatically always force an exact match behind the scenes unless it found circumflex at the beginning or dollar sign at the end, which would accomplish the same thing.

2. Match a programme whose name consists entirely of digits, e.g., `^1864$`, which is very rare. This could no doubt be done automatically by GiA by detecting names consisting of digits. Even that would be unnecessary if exact matches were done by default as described above.

3. Generalise a programme name match to catch multiple series, e.g., going from `The Repair Shop: Series 1` to `^The Repair Shop: Series` or `^The Repair Shop`.  Programme names are unique enough that the circumflex would virtually never be needed, but if GiA respected it as described above, users would have that flexibility.

Regular expression matches also work differently when searching than when adding series-link items to the queue. For example, you need `^Click [^-]` to accomplish in a search what `^Click$` does when adding series-link items (though there would never be a real need to do that). `^606` would work in a search, but `^606$` would fail even though it works for series-link. Nobody is going to understand the difference unless they know how GiA interfaces with get_iplayer  and how get_iplayer searching works. 

Taken together, the above makes me think it would be unnecessary cognitive load to surface regular expressions in the PVR. If you know you can use them, you might come across a novel reason to do so. Most users, with a little help from GiA, will never need to use them.

**Comment by sleaux-meaux** [comment]
I think there is a slightly simpler solution. Rather than doing an exact match, GiA could automatically force a match at the beginning of the programme name, i.e., add a circumflex at the beginning of the PVR show name unless one has been put there by the user. In an exceptionally rare case like Click, where the match should be truncated, users could still add a dollar sign. Previous post edited accordingly.
